gcsio/src/main/java/com/google/cloud/hadoop/gcsio/Watchdog.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.hadoop.gcsio;
+
+import com.google.api.core.ApiClock;
+import com.google.common.base.Preconditions;
+import com.google.common.flogger.GoogleLogger;
+import io.grpc.ClientCall;
+import io.grpc.Context.CancellableContext;
+import io.grpc.stub.StreamObserver;
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Prevents the streams from hanging indefinitely. This middleware garbage collects idle streams in
+ * case the user forgot to close a ServerStream or if a connection is reset and GRPC does not get
+ * notified.
+ *
+ * <p>For every {@code checkInterval}, this class checks thresholds:
+ *
+ * <ul>
+ *   <li>idleTimeout: the amount of time to wait before assuming that the caller forgot to close the
+ *       stream and forcefully closing the stream. This is measured from the last time the caller
+ *       had no outstanding demand. Duration.ZERO disables the timeout.
+ * </ul>
+ */
+public final class Watchdog implements Runnable {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  // Dummy value to convert the ConcurrentHashMap into a Set
+  private static final Object PRESENT = new Object();
```

Comment:
Can you use [`ConcurrentHashMap.newKeySet()`](https://stackoverflow.com/a/6992643/3227693) instead?